File: packages/mcp/src/manifest.ts
import { readdirSync, readFileSync, statSync } from "node:fs";
import { join, resolve } from "node:path";

export type WidgetEntry = {
  name: string;
  html: string;
};

export type WidgetManifest = {
  baseUrl: string;
  widgets: Record<string, WidgetEntry>;
};

function toPosix(path: string): string {
  return path.replaceAll("\\", "/");
}

function normalizeBaseUrl(baseUrl: string): string {
  return baseUrl.replace(/\/+$/, "");
}

function absolutizeAssetUrls(html: string, baseUrl: string): string {
  const base = normalizeBaseUrl(baseUrl);
  return html
    .replace(/\b(src|href)=["'](\/?assets\/[^"']+)["']/g, (_match, attr: string, path: string) => {
      const normalized = path.startsWith("/") ? path : `/${path}`;
      return `${attr}="${base}${normalized}"`;
    })
    .replace(/\b(src|href)=["']\.\/assets\/([^"']+)["']/g, (_match, attr: string, path: string) => {
      return `${attr}="${base}/assets/${path}"`;
    });
}

function collectWidgetHtmlFiles(widgetsDir: string): Array<{ name: string; filePath: string }> {
  let entries: string[];
  try {
    entries = readdirSync(widgetsDir);
  } catch {
    return [];
  }

  const widgets: Array<{ name: string; filePath: string }> = [];
  for (const entry of entries) {
    const widgetDir = join(widgetsDir, entry);
    try {
      if (!statSync(widgetDir).isDirectory()) {
        continue;
      }
    } catch {
      continue;
    }
    const htmlPath = join(widgetDir, "index.html");
    try {
      if (statSync(htmlPath).isFile()) {
        widgets.push({ name: entry, filePath: htmlPath });
      }
    } catch {
      // missing index.html
    }
  }
  return widgets;
}

export function loadWidgetManifest(projectRoot: string, baseUrl: string): WidgetManifest {
  const root = resolve(projectRoot);
  const widgetsDir = join(root, "dist", "widgets");
  const discovered = collectWidgetHtmlFiles(widgetsDir);
  if (discovered.length === 0) {
    throw new Error(
      `No widget HTML found under ${toPosix(widgetsDir)}. Run vite build with the belgie() plugin first.`,
    );
  }

  const widgets: Record<string, WidgetEntry> = {};
  for (const widget of discovered) {
    const html = absolutizeAssetUrls(readFileSync(widget.filePath, "utf-8"), baseUrl);
    widgets[widget.name] = {
      name: widget.name,
      html,
    };
  }

  return {
    baseUrl: normalizeBaseUrl(baseUrl),
    widgets,
  };
}

File: src/belgie/mcp/_manifest.py
from pathlib import Path
from re import Match, Pattern, compile as compile_pattern
from typing import Final
from urllib.parse import urljoin, urlparse

from pydantic import BaseModel, ConfigDict, Field

ASSET_URL_PATTERN: Final[Pattern[str]] = compile_pattern(
    r'\b(src|href)=["\']((?:\.?/)?assets/[^"\']+)["\']',
)
INVALID_BASE_URL_ERROR: Final[str] = "base_url must be an absolute http(s) URL, got {base_url!r}"
MISSING_WIDGET_HTML_ERROR: Final[str] = "No widget HTML found under {widgets_dir}. Run `belgie run vite build` first."


class WidgetEntry(BaseModel):
    model_config = ConfigDict(frozen=True, populate_by_name=True)

    name: str
    html: str


class WidgetManifest(BaseModel):
    model_config = ConfigDict(frozen=True, populate_by_name=True)

    base_url: str = Field(validation_alias="baseUrl")
    widgets: dict[str, WidgetEntry]


def load_widget_manifest(*, base_url: str, project_path: Path | None = None) -> WidgetManifest:
    normalized_base_url = normalize_base_url(base_url)
    resolved_project_path = resolve_project_path(project_path)
    widgets_dir = resolved_project_path / "dist" / "widgets"
    widgets: dict[str, WidgetEntry] = {}
    if widgets_dir.is_dir():
        for widget_dir in sorted(widgets_dir.iterdir()):
            html_path = widget_dir / "index.html"
            if not widget_dir.is_dir() or not html_path.is_file():
                continue
            widgets[widget_dir.name] = WidgetEntry(
                name=widget_dir.name,
                html=absolutize_asset_urls(html_path.read_text(encoding="utf-8"), normalized_base_url),
            )
    if not widgets:
        msg = MISSING_WIDGET_HTML_ERROR.format(widgets_dir=widgets_dir)
        raise FileNotFoundError(msg)
    return WidgetManifest(base_url=normalized_base_url, widgets=widgets)


def absolutize_asset_urls(html: str, base_url: str) -> str:
    def replace(match: Match[str]) -> str:
        attribute, asset_path = match.groups()
        normalized_path = asset_path.removeprefix("./").removeprefix("/")
        return f'{attribute}="{urljoin(f"{base_url}/", normalized_path)}"'

    return ASSET_URL_PATTERN.sub(replace, html)


def normalize_base_url(base_url: str) -> str:
    parsed = urlparse(base_url)
    if parsed.scheme not in {"http", "https"} or not parsed.netloc:
        msg = INVALID_BASE_URL_ERROR.format(base_url=base_url)
        raise ValueError(msg)
    return base_url.rstrip("/")


def resolve_project_path(path: Path | None) -> Path:
    return (Path.cwd() if path is None else Path(path)).resolve()
